File: tests/regress/emu_clear_errors.py
import regress
from unicorn import *
from unicorn.x86_const import *

CODE = (
    b'\x8B\x74\x01\x28'  # mov esi, dword ptr [ecx + eax + 0x28]  mapped: 0x1000
    b'\x03\xF0'          # add esi, eax                                   0x1004
    b'\x8D\x45\xFC'      # lea eax, dword ptr [ebp - 4]                   0x1006
    b'\x50'              # push eax                                       0x1009
    b'\x6A\x40'          # push 0x40                                      0x100A
    b'\x6A\x10'          # push 0x10                                      0x100C
    b'\x56'              # push esi                                       0x100E
)
BASE = 0x1000
STACK = 0x4000


class EmuClearErrorsTest(regress.RegressTest):

    def test_hook_code_stop_emu(self):
        mu = Uc(UC_ARCH_X86, UC_MODE_32)

        # base of CODE
        mu.mem_map(BASE, 0x1000)
        mu.mem_write(BASE, CODE)

        # base of STACK
        mu.mem_map(STACK, 0x1000)
        mu.mem_write(STACK, b"\x00" * 0x1000)

        mu.reg_write(UC_X86_REG_EIP, BASE)
        mu.reg_write(UC_X86_REG_ESP, STACK + 0x1000 - 8)
        mu.reg_write(UC_X86_REG_EBP, STACK + 0x1000 - 8)
        mu.reg_write(UC_X86_REG_ECX, 0x0)
        mu.reg_write(UC_X86_REG_EAX, 0x0)

        # we only expect the following instruction to execute,
        #  and it will fail, because it accesses unmapped memory.
        # mov esi, dword ptr [ecx + eax + 0x28]    mapped: 0x1000

        with self.assertRaises(UcError) as ex:
            mu.emu_start(BASE, BASE + len(CODE), count=1)

        self.assertEqual(UC_ERR_READ_UNMAPPED, ex.exception.errno)

        regress.logger.debug("pc: %#x", mu.reg_read(UC_X86_REG_EIP))

        # now, we want to reuse the emulator, and keep executing
        #  from the next instruction

        # we expect the following instructions to execute
        #   add esi, eax                                   0x1004
        #   lea eax, dword ptr [ebp - 4]                   0x1006
        #   push eax                                       0x1009
        #   push 0x40                                      0x100A
        #   push 0x10                                      0x100C
        #   push esi                                       0x100E
        mu.emu_start(BASE + 0x4, BASE + len(CODE))

        regress.logger.debug("pc: %#x", mu.reg_read(UC_X86_REG_EIP))


if __name__ == '__main__':
    regress.main()

File: tests/regress/emu_clear_errors.c
#include <sys/types.h>
#include <stdlib.h>
#include <string.h>

#include <unicorn/unicorn.h>

static int count = 1;

bool cb_hookunmapped(uc_engine *uc, uc_mem_type type, uint64_t address, uint32_t size, int64_t value, void *user_data) {
   uint32_t pc = 0;
   uc_reg_read(uc, UC_X86_REG_EIP, &pc);
   fprintf(stderr, "mem unmapped: 0x%x type: %x address: 0x%"PRIx64" length: %x value: 0x%"PRIx64"\n", 
           pc, type, address, size, value);

   uc_err err = UC_ERR_OK;
   err = uc_emu_stop(uc);
   if (err != UC_ERR_OK) {
       fprintf(stderr, "stop not ok");
       exit(0);
   }
   return true;
}

// move esi, dword ptr [ecx + eax + 0x28]
// add esi, eax
// lea eax, dword ptr [ebp - 4]
// push eax
// push 0x40
// push 0x10
// push esi
// call some address
#define CODE "\x8B\x74\x01\x28" \
             "\x0C\xF0" \
             "\x8D\x45\xFC" \
             "\x50" \
             "\x6A\x40" \
             "\x6A\x10" \
             "\x56" \
             "\xFF\x15\x20\x20\x00\x10"

int main() {
   uc_engine *uc;

   uc_err err = uc_open(UC_ARCH_X86, UC_MODE_32, &uc);
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_open\n", count++);

   err = uc_mem_map(uc, 0x1000, 0x1000, UC_PROT_ALL);
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_mem_map: code\n", count++);

   uint8_t code[0x1000];
   memset(code, 0x0, sizeof(code));
   memcpy(code, CODE, sizeof(CODE));

   err = uc_mem_write(uc, 0x1000, code, sizeof(code));
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_mem_write: code\n", count++);

   uint32_t eip = 0x1000;
   err = uc_reg_write(uc, UC_X86_REG_EIP, &eip);
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_reg_write: eip\n", count++);

   err = uc_mem_map(uc, 0x4000, 0x4000, UC_PROT_ALL);
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_mem_map: stack\n", count++);

   uint8_t stack[0x4000];
   memset(stack, 0x0, sizeof(stack));

   err = uc_mem_write(uc, 0x4000, code, sizeof(code));
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_mem_write: stack\n", count++);

   uint32_t esp = 0x6000;
   err = uc_reg_write(uc, UC_X86_REG_ESP, &esp);
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_reg_write: esp\n", count++);

   uint32_t ebp = 0x6000;
   err = uc_reg_write(uc, UC_X86_REG_EBP, &ebp);
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_reg_write: ebp\n", count++);

   uc_hook h1;

   err = uc_hook_add(uc, &h1, UC_HOOK_MEM_UNMAPPED, cb_hookunmapped, NULL, 1, 0);
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_hook_add\n", count++);

   // this should execute only a single instruction at 0x1000, because
   // that instruction accesses invalid memory.
   err = uc_emu_start(uc, 0x1000, 0x100F, 0, 0);
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_emu_start\n", count++);

   // yes, not necessary, but to demonstrate the UC API is working as expected
   eip = 0x1004;
   err = uc_reg_write(uc, UC_X86_REG_EIP, &eip);
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_reg_write: eip\n", count++);

   // this should execute the remaining instructions up to (but not includign) 0x100F.
   // currently, it returns an error about an unmapped read.
   // seems that this error should have been returned in the previous call
   //  to emu_start.
   err = uc_emu_start(uc, 0x1004, 0x100F, 0, 0);
   if (err != UC_ERR_OK) {
      fprintf(stderr, "not ok %d - %s\n", count++, uc_strerror(err));
      exit(0);
   }
   fprintf(stderr, "ok %d - uc_emu_start\n", count++);

   fprintf(stderr, "ok %d - Done", count++);

   return 0;
}
